[WASM-Function-References] Extend tables with initializer expressions

takikawa opened the pull request:
#### 42927fcd77eacfe0e121f8ec1519d22dc8344468
<pre>
[WASM-Function-References] Extend tables with initializer expressions
<a href="https://bugs.webkit.org/show_bug.cgi?id=251123">https://bugs.webkit.org/show_bug.cgi?id=251123</a>

Reviewed by Justin Michaud.

Add support for tables with initializer expressions. This entails extending
TableInformation with initialization information, just as there is for globals.
The same fast paths for initializer expressions are used for tables, except
restricted to the ones that make sense for reference types.

* JSTests/wasm/funcref-spec-harness/sync_index.js:
* JSTests/wasm/function-references-spec-tests/table.wast.js: Added.
* JSTests/wasm/function-references/table_init.js: Added.
(module):
(async testTableInitParsing):
(async testTableInitRuntime):
* JSTests/wasm/gc/table_init.js: Added.
(testTableValidation):
* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::TableInformation::TableInformation):
(JSC::Wasm::TableInformation::initType const):
(JSC::Wasm::TableInformation::initialBitsOrImportNumber const):
* Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parseTableHelper):
* Source/JavaScriptCore/wasm/js/WebAssemblyModuleRecord.cpp:
(JSC::WebAssemblyModuleRecord::initializeExports):

Canonical link: <a href="https://commits.webkit.org/269527@main">https://commits.webkit.org/269527@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/44f411124d2f17ff5caba829087976e81dd7208e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22770 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/431 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23857 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24678 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21075 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/1504 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23296 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21984 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23010 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/267 "Found 2 new test failures: http/tests/download/area-download.html, http/tests/site-isolation/draw-after-navigation.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19747 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25531 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/275 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20627 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26840 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/19822 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20636 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20871 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24695 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/22132 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/313 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18141 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/26189 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/243 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6134 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5437 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/368 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/27471 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/306 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5963 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/wasm/js/WebAssemblyModuleRecord.cpp:
```diff
@@ -534,6 +513,58 @@ void WebAssemblyModuleRecord::initializeExports(JSGlobalObject* globalObject)
+            scope.assertNoException();
+            ASSERT(table);
+            m_instance->setTable(vm, i, table);
+            RETURN_IF_EXCEPTION(scope, void());
```
Should we crash instead?

takikawa replied:
I looked into this and I think it's fine to not crash here, since the exception is checked by the callers of `initializeExports` and should error out soon after.

justinmichaud reviewed (approved):
r=me

takikawa committed:
[WASM-Function-References] Extend tables with initializer expressions
https://bugs.webkit.org/show_bug.cgi?id=251123

Reviewed by Justin Michaud.

Add support for tables with initializer expressions. This entails extending
TableInformation with initialization information, just as there is for globals.
The same fast paths for initializer expressions are used for tables, except
restricted to the ones that make sense for reference types.

* JSTests/wasm/funcref-spec-harness/sync_index.js:
* JSTests/wasm/function-references-spec-tests/table.wast.js: Added.
* JSTests/wasm/function-references/table_init.js: Added.
(module):
(async testTableInitParsing):
(async testTableInitRuntime):
* JSTests/wasm/gc/table_init.js: Added.
(testTableValidation):
* Source/JavaScriptCore/wasm/WasmFormat.h:
(JSC::Wasm::TableInformation::TableInformation):
(JSC::Wasm::TableInformation::initType const):
(JSC::Wasm::TableInformation::initialBitsOrImportNumber const):
* Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
* Source/JavaScriptCore/wasm/WasmSectionParser.cpp:
(JSC::Wasm::SectionParser::parseTableHelper):
* Source/JavaScriptCore/wasm/js/WebAssemblyModuleRecord.cpp:
(JSC::WebAssemblyModuleRecord::initializeExports):

Canonical link: https://commits.webkit.org/269527@main
